File: DSA/Stack/Code/InfixToPrefix.py
# Infix to Prefix in Python

def precedence(c):
    if c in ['+', '-']:
        return 1
    if c in ['*', '/']:
        return 2
    if c == '^':
        return 3
    return 0

def is_operator(c):
    return c in ['+', '-', '*', '/', '^']

def infix_to_prefix(infix):
    # Reverse the string
    infix = infix[::-1]
    
    # Swap ( and )
    infix = infix.replace('(', 'TEMP').replace(')', '(').replace('TEMP', ')')
    
    # Convert to postfix
    stack = []
    postfix = ""
    
    for c in infix:
        if c.isalnum():
            postfix += c
        elif c == '(':
            stack.append(c)
        elif c == ')':
            while stack and stack[-1] != '(':
                postfix += stack.pop()
            stack.pop()
        elif is_operator(c):
            while stack and precedence(stack[-1]) >= precedence(c):
                postfix += stack.pop()
            stack.append(c)
    
    while stack:
        postfix += stack.pop()
    
    # Reverse result
    return postfix[::-1]

# Test
infix = "a+b*c-d/e"

print("=== Infix to Prefix ===\n")
print("Algorithm: Reverse -> Convert to Postfix -> Reverse")
print("Steps:")
print("1. Reverse the infix expression")
print("2. Swap ( and )")
print("3. Convert to postfix")
print("4. Reverse result\n")

print(f"Infix: {infix}")
print(f"Prefix: {infix_to_prefix(infix)}\n")
print("Complexity: O(n)")

File: DSA/Stack/Code/InfixToPrefix.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <stack>
using namespace std;

// Infix to Prefix in C++
int precedence(char c) {
    if (c == '+' || c == '-') return 1;
    if (c == '*' || c == '/') return 2;
    if (c == '^') return 3;
    return 0;
}

bool isOperator(char c) {
    return (c == '+' || c == '-' || c == '*' || c == '/' || c == '^');
}

string infixToPrefix(string infix) {
    // Reverse the string
    reverse(infix.begin(), infix.end());
    
    // Swap ( and )
    for (char& c : infix) {
        if (c == '(') c = ')';
        else if (c == ')') c = '(';
    }
    
    // Convert to postfix
    stack<char> s;
    string postfix = "";
    
    for (char c : infix) {
        if (isalnum(c)) {
            postfix += c;
        } else if (c == '(') {
            s.push(c);
        } else if (c == ')') {
            while (!s.empty() && s.top() != '(') {
                postfix += s.top();
                s.pop();
            }
            s.pop();
        } else if (isOperator(c)) {
            while (!s.empty() && precedence(s.top()) >= precedence(c)) {
                postfix += s.top();
                s.pop();
            }
            s.push(c);
        }
    }
    
    while (!s.empty()) {
        postfix += s.top();
        s.pop();
    }
    
    // Reverse result
    reverse(postfix.begin(), postfix.end());
    return postfix;
}

int main() {
    cout << "=== Infix to Prefix ===\n\n";
    cout << "Algorithm: Reverse -> Convert to Postfix -> Reverse\n";
    cout << "Steps:\n";
    cout << "1. Reverse the infix expression\n";
    cout << "2. Swap ( and )\n";
    cout << "3. Convert to postfix\n";
    cout << "4. Reverse result\n\n";
    
    string infix = "a+b*c-d/e";
    cout << "Infix: " << infix << "\n";
    cout << "Prefix: " << infixToPrefix(infix) << "\n\n";
    cout << "Complexity: O(n)\n";
    
    return 0;
}
